### PR TITLE
Feature/derive key from password

### PR DESCRIPTION
**Pull request by stnmtz**
- Fixes broken deriveKeyFromMaster method and replaces it with deriveKeyFromPassword
- Tests it ;)

### CODE REVIEW

**Comment by stnmtz** [comment]
I did not do a major bump, as the former code was broken and nobody could have used it. I consider it more a deletion of unused code and adding a feature.

**Comment by jkoenig134** [comment]
alright.

File: package.json [hunk 1]
```diff
@@ -44,25 +44,26 @@
...
         "@js-soft/license-check": "1.0.9",
         "@js-soft/ts-serval": "2.0.10",
         "@types/chai": "^5.0.0",
         "@types/libsodium-wrappers-sumo": "^0.7.8",
-        "@types/mocha": "^10.0.8",
-        "@types/node": "^22.7.4",
+        "@types/mocha": "^10.0.9",
+        "@types/node": "^22.7.7",
         "@types/uuid": "^10.0.0",
+        "@typescript-eslint/eslint-plugin": "^8.10.0",
```
Comment (jkoenig134) [inline package.json:55]: can you remove this please? The correct version of this lib is brought by @js-soft/eslint-config-ts

File: src/CryptoDerivation.ts [hunk 2]
```diff
@@ -30,30 +33,65 @@ export interface ICryptoDerivationStatic {
...
     ): Promise<CryptoSecretKey> {
+        const sodium: any = (await SodiumWrapper.ready()) as any;
+        if (salt.buffer.byteLength !== 16) {
+            throw new Error(`The salt must be exactly ${sodium.crypto_pwhash_SALTBYTES} bytes long!`);
+        }
+        if (opslimit < sodium.crypto_pwhash_OPSLIMIT_MIN) {
+            throw new Error(`The opslimit must be higher than ${sodium.crypto_pwhash_OPSLIMIT_MIN}.`);
+        }
+        if (sodium.crypto_pwhash_OPSLIMIT_MAX > 0 && opslimit > sodium.crypto_pwhash_OPSLIMIT_MAX) {
+            throw new Error(`The opslimit must be lower than ${sodium.crypto_pwhash_OPSLIMIT_MAX}.`);
```
Comment (Magnus-Kuhn) [inline src/CryptoDerivation.ts:52]: How can that constant not be larger than 0?

File: test/crypto/CryptoDerivationTest.test.ts [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+export class CryptoDerivationTest {
+    public static run(): void {
+        describe("CryptoDerivation", function () {
+            describe("Execute deriveKeyFromBase()", function () {
+                let keybuffer: ICoreBuffer;
+                before(async function () {
+                    // Create 256bit entropy
+                    keybuffer = await CryptoHash.hash(
+                        CoreBuffer.fromString("test", Encoding.Utf8),
```
Comment (Magnus-Kuhn) [inline test/crypto/CryptoDerivationTest.test.ts:22]: `.fromUtf8` is possible in a couple of spots

File: test/crypto/CryptoDerivationTest.test.ts [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+                        "eyJrZXkiOiJfdkg4NkdJeU12cFdnZkJ2Y09FY0dfNHRhakpiVXJ5cGRUWW9iZl9jb0Y4IiwiYWxnIjozLCJAdHlwZSI6IkNyeXB0b1NlY3JldEtleSJ9"
+                    );
+
+                    const derivedComparison = await CryptoDerivation.deriveKeyFromBase(keybuffer, 0, "12345678");
+                    expect(derivedComparison).to.be.instanceOf(CryptoSecretKey);
+                    expect(derivedComparison.toBase64()).to.equal(
+                        "eyJrZXkiOiJfdkg4NkdJeU12cFdnZkJ2Y09FY0dfNHRhakpiVXJ5cGRUWW9iZl9jb0Y4IiwiYWxnIjozLCJAdHlwZSI6IkNyeXB0b1NlY3JldEtleSJ9"
+                    );
+
+                    expect(derived.toBase64()).to.equal(derivedComparison.toBase64());
```
Comment (Magnus-Kuhn) [inline test/crypto/CryptoDerivationTest.test.ts:40]: Those equality checks (and later inequality checks) are redundant

File: src/CryptoDerivation.ts (hunk 2 above)
Comment (stnmtz) [reply src/CryptoDerivation.ts:54]: The default was -1 which means "no limit"

**Review by jkoenig134 — DISMISSED** [review]

File: test/crypto/CryptoDerivationTest.test.ts (hunk 4 above)
Comment (stnmtz) [reply test/crypto/CryptoDerivationTest.test.ts:36]: I've removed the redundant string equality checks. I left the not equal test as they are, as I would like to see if anything changes here in future versions (which shouldn't be the case). If that is the case, we know which derived value did change (and possibly why) and not only that they are not equal.

**Review by jkoenig134 — APPROVED** [review]

**Pull request merged**